Support `__atomic_compare_exchange`

maul-esel opened the pull request:
This PR adds support for the builtin atomic functions `__atomic_compare_exchange_n` and `__atomic_compare_exchange`.

maul-esel committed:
first implementation of __atomic_compare_exchange_n

maul-esel committed:
add support for __atomic_compare_exchange

maul-esel committed:
avoid wrapping inner memory order check in atomic block

While likely not a soundness problem, this is problematic due to the
LBE performed by CfgBuilder which fuses the overapproximation assert
with other statements, such that the entire block is marked as over-
approximation. Thus we report UNKNOWN unnecessarily.

maul-esel committed:
add regression tests for __atomic_compare_exchange(_n)

schuessf commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/standardfunctions/StandardFunctionHandler.java:
```diff
@@ -1395,30 +1550,35 @@ private Result handleAtomicFetch(final IDispatcher main, final IASTFunctionCallE
 	 */
 	private ExpressionResult applyMemoryOrder(final ILocation loc, final ExpressionResult body,
-			final Expression memoryOrder) {
+			final Expression... memoryOrders) {
```
You updated the signature of the function, can you also update the documentation? It is not quite clear to me, how the multiple memory orders are handled here.

maul-esel replied:
:+1: hopefully clearer now

schuessf replied:
Thanks!

> If every given memory order is equal to MEMORY_ORDER_SEQ_CST, we just make all statements atomic.

If understand correctly, you create nested ifs:

```
if (mem1 == SEQ_CST) {
  if (mem2 ==  SEQ_CST) {
    atomic { st };
  } else {
    assert false;
} else {
  assert false;
}
```

Wouldn't it be simpler to create simply the conjunction of all memory-order equalities?

```
if (mem1 == SEQ_CST && mem2 ==  SEQ_CST) {
    atomic { st };
} else {
  assert false;
}
```

schuessf commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/standardfunctions/StandardFunctionHandler.java:
```diff
@@ -1353,6 +1430,84 @@ private Result handleAtomicExchangeN(final IDispatcher main, final IASTFunctionC
+	//
+	// @formatter:off
+	// (evaluate arguments)
+	// atomic {
```
Thanks for the illustration. This really helps to understand the specification.

schuessf commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/standardfunctions/StandardFunctionHandler.java:
```diff
@@ -1353,6 +1430,84 @@ private Result handleAtomicExchangeN(final IDispatcher main, final IASTFunctionC
+		final var atomicBody = new ExpressionResultBuilder().addAuxVarWithDeclaration(success)
+				.addStatement(StatementFactory.constructIfStatement(loc, weakResult.getLrValue().getValue(),
+						new Statement[] { new HavocStatement(loc, new VariableLHS[] { success.getLhs() }) },
+						new Statement[] { StatementFactory.constructAssignmentStatement(loc,
```
There is also the method `StatementFactory::constructSingleAssignmentStatement`, which improves readability (it already creates singleton arrays).

schuessf commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/standardfunctions/StandardFunctionHandler.java:
```diff
@@ -1353,6 +1430,84 @@ private Result handleAtomicExchangeN(final IDispatcher main, final IASTFunctionC
+	//
+	// @formatter:off
+	// (evaluate arguments)
+	// atomic {
```
Is this block really atomic or does it depend on the memory order?

maul-esel replied:
I've added a comment explaining that we only follow this scheme if the memory order is sequential consistency, and otherwise `assert false`.

schuessf commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/standardfunctions/StandardFunctionHandler.java:
```diff
@@ -1353,6 +1430,84 @@ private Result handleAtomicExchangeN(final IDispatcher main, final IASTFunctionC
+		final var successBoolean =
+				mExpressionTranslation.constructBinaryEqualityExpression(loc, IASTBinaryExpression.op_notequals,
+						success.getExp(), boolType, mExpressionTranslation.constructZero(loc, boolType), boolType);
+		final var atomicBody = new ExpressionResultBuilder().addAuxVarWithDeclaration(success)
```
This is a huge single statement. I think, it should improve readability if you refactor it in smaller statements.

maul-esel replied:
I found it sometimes difficult to read other translation code if it first assembles lots of little pieces of Boogie (without context what they are useful for) and then assembles them at the end in some order.

So I tried to stick as closely as possible to the structure of the generated Boogie code here... though I can't say the result is necessarily much better :face_with_diagonal_mouth:

schuessf replied:
Okay, I see. I think it looks better now. Maybe can improve the readability further to move the constructor call (maybe with the aux var) and the `build()` to separate statements. I think the indentation could be reduced in that case.

schuessf commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/standardfunctions/StandardFunctionHandler.java:
```diff
@@ -1305,6 +1302,86 @@ private Result handleAtomicExchange(final IDispatcher main, final IASTFunctionCa
+	//   write(ptr_val, expected)
+	// return success
+	// @formatter:on
+	private Result handleAtomicCompareExchange(final IDispatcher main, final IASTFunctionCallExpression node,
```
If I understand the specification correctly, then the only difference between `__atomic_compare_exchange_n` and `__atomic_compare_exchange` is that the latter one has a pointer as `desired` that is simply dereferenced and written into `ptr`. Is it possible to refactor this method to call `handleAtomicCompareExchangeN`?

maul-esel replied:
:+1: done

schuessf commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/standardfunctions/StandardFunctionHandler.java:
```diff
@@ -1353,6 +1430,84 @@ private Result handleAtomicExchangeN(final IDispatcher main, final IASTFunctionC
+		final var expectedWrite = mExprResultTransformer.makePointerAssignment(loc, expectedResult.getLrValue(),
+				pointerRead.getLrValue().getValue());
+		return resultBuilder.addAllExceptLrValue(atomic).addAllExceptLrValueAndStatements(expectedWrite)
+				.addStatement(StatementFactory.constructIfStatement(loc, successBoolean, new Statement[0],
```
Minor note: Is there any reason why you create `if (success) {} else st` instead of `if (!success) st` here? It should not make a difference on the CFG, but the Boogie code of the latter looks slightly better?

maul-esel replied:
I forgot about that. I didn't see how to apply negation at first, and only wanted to test out the method briefly :D good catch, fixed now.

schuessf reviewed (commented):
Thanks for the PR! I think it looks good, I have just some minor comments.

maul-esel committed:
StandardFunctionHandler: update name and documentation of applyMemoryOrder

maul-esel committed:
use utility method constructSingleAssignmentStatement

maul-esel committed:
clarify role of schema wrt memory order

maul-esel committed:
reduce duplication between handleAtomicCompareExchange and handleAtomicCompareExchangeN

maul-esel committed:
handleAtomicCompareExchange: slightly simplify generated Boogie code

schuessf commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/standardfunctions/StandardFunctionHandler.java:
```diff
@@ -1301,10 +1298,123 @@ private Result handleAtomicExchange(final IDispatcher main, final IASTFunctionCa
+
+		final var expectedWrite = mExprResultTransformer.makePointerAssignment(loc, expectedResult.getLrValue(),
+				pointerRead.getLrValue().getValue());
+		return resultBuilder.addAllExceptLrValue(atomic).addAllExceptLrValueAndStatements(expectedWrite)
```
Is it correct that the write is indeed not atomic? I don't fully understand this from the specification.

maul-esel replied:
Hmm, I'm not sure. I was basing this on the description at https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html#index-_005f_005fatomic_005fcompare_005fexchange_005fn which says:

> If equal, the operation is a _read-modify-write_ operation that writes `desired` into `*ptr`. If they are not equal, the operation is a _read_ and the current contents of `*ptr` are written into `*expected.`

I understood that to say that the write is at least not atomic with the rest, otherwise the operation would be a _read-modify-write_. Though that still leaves open whether the write is atomic by itself.

The description of `atomic_compare_exchange` in Section 7.17.7.4 of the C standard sounds somewhat different: 
![image](https://github.com/user-attachments/assets/b4f74ca7-d6d8-4fc3-a20a-249fa502608d)

This seems more like the failure case is atomic with the rest. Of course, we're implementing the GNU variant, which "approximates" the standard... :shrug:

maul-esel replied:
The C standard is also a bit more precise what "fail spuriously" looks like; it's different from the behaviour for different values. I will have to update the code for that as well.
![image](https://github.com/user-attachments/assets/b44845a8-1a35-45f0-bdc1-6060ae2bde20)

maul-esel replied:
I've reimplemented the method based on the C standard description.

schuessf commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/standardfunctions/StandardFunctionHandler.java:
```diff
@@ -1301,10 +1298,123 @@ private Result handleAtomicExchange(final IDispatcher main, final IASTFunctionCa
+						DataStructureUtils.concat(desiredRead.getStatements(), pointerWrite.getStatements())
+								.toArray(Statement[]::new),
+						new Statement[] { StatementFactory.constructSingleAssignmentStatement(loc, success.getLhs(),
+								mExpressionTranslation.constructLiteralForIntegerType(loc, boolType,
```
You could use `boolZero` here

maul-esel committed:
simplify Boogie code for applyMemoryOrders, take advantage of ExpressionFactory for built-in simplication

maul-esel committed:
Reimplement atomic_compare_exchange

Follow description in C standard rather than unclear cppref / GNU
descriptions (the C standard is still pretty vague though).
Introduce some auxiliary methods to make the code easier to read.
Optimize code when "weak" argument is a literal and adapt a program
to test this.

schuessf reviewed (approved):
Looks good! I just assume your implementation follows the vague standard 😉
